update_blocklist.nix

```
@@ -40,16 +40,44 @@ in
   # Create an ip set and add each ip to it one by one
 
   # IPv4 and IPv6 regex patterns with CIDR notation support - WARNING: might not be correct for all IPs (e.g. ignore valid ones or accept wrong ones), but seems to work fine
+  # Also supports hosts files, as well as plain lists of domains.
   ipv4_regex="^([0-9]{1,3}\.){3}[0-9]{1,3}(\/[0-9]{1,2})?$"
   ipv6_regex="^([0-9a-fA-F:]+::?[0-9a-fA-F]*)+(\/[0-9]{1,3})?$"
+  domain_regex"^([A-Za-z0-9-]+(?<!-)\.)+[A-Za-z0-9-]+$"
+  host_regex"$^0\.0\.0\.0 $\{domain_regex:1\}"
+
+  # Define the functions to be used.
+  blockIPv4 () {
+      echo "${ipSetName}" "$1"
+  }
+
+  blockIPv6 () {
+      echo "${ipV6SetName}" "$1"
+  }
+
+  blockDomain () {
+      hostIPv4=$(dig $1 A +short)
+      while IFS= read -r IP; do
+          echo $(blockIPv4 $1)
+      done < $hostIPv4s
+
+      hostIPv6=$(dig $1 AAAA +short)
+      while IFS= read -r IP; do
+          echo $(blockIPv6 $1)
+      done < $hostIPv6s
+  }
 
   # Use a temporary buffer to improve performance
   {
       while IFS= read -r IP; do
           if [[ $IP =~ $ipv4_regex ]]; then
-              echo -exist add "${ipSetName}" "$IP"
-          elif [[ $IP =~ $ipv6_regex ]]; then
-              echo -exist add "${ipV6SetName}" "$IP"
+              blockIPv4 $IP
+          if [[ $IP =~ $ipv6_regex ]]; then
+              blockIPv6 $IP
+          elif [[ $IP =~ $host_regex ]]; then
+              blockDomain $\{IP:8\}
```

Is that the correct syntax for escaping the `${` from nix? I thought it would be
```suggestion
              blockDomain \${IP:8}
```
🤔

---

Hmm.. Looking it up, I think maybe we're both wrong: see [this](https://discourse.nixos.org/t/dealing-with-shell-variables-in-nix-strings/7978/2) NixOS discourse page, and [this](https://www.reddit.com/r/NixOS/comments/jmlohf/escaping_interpolation_in_bash_string/) Reddit page.

Thanks for pointing this out!

---

Right - that would also work! Apparently the backslash should as well: https://nix.dev/manual/nix/2.26/language/string-literals#string-literals - I don't have a preference. Choose one!

---

Alright, the backslash is more common in other languages, so that might be more recognizable.

I'll go with that!

---

That didn't work! `''` would make more sense, then.

---

Interesting… I thought the documentation would allow it… Yeah - then let's go with `''` 🤷